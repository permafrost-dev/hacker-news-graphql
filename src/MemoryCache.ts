export interface CacheItem {
    ts: number;
    expires: number;
    value: any;
}

export interface CacheMap {
    [key: string]: CacheItem;
}

export class MemoryCache {
    public map: CacheMap = {};

    public has(key: string): boolean {
        this.purge();

        return typeof this.map[key] !== 'undefined';
    }

    public get(key: string): any {
        this.purge();

        return this.map[key]?.value ?? null;
    }

    public put(key: string, value: any, ttlSeconds: number) {
        this.map[key] = {
            ts: new Date().getTime(),
            expires: new Date().getTime() + ttlSeconds * 1000,
            value: value,
        };
    }

    public purge() {
        for (const prop in this.map) {
            if (new Date().getTime() >= this.map[prop].expires) {
                delete this.map[prop];
                continue;
            }
        }
    }
}
