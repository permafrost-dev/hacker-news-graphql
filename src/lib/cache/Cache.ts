/* eslint-disable no-unused-vars */

export abstract class Cache {
    public prefix = '';

    public destroy() {
        //
    }
    public abstract has(key: string): Promise<boolean>;

    public abstract get(key: string): Promise<any | null>;

    public abstract put(key: string, value: any, ttlSeconds: number): void;

    public abstract purge();

    public abstract clear(): boolean;
}
