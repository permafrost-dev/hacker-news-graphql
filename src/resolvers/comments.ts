/* eslint-disable no-unused-vars */

import axios from 'axios';

export function getResolver(cache: any): any {
    const commentsResolver = async (parent, { first }) => {
        const comments: Record<string, any>[] = [];

        if (typeof parent['kids'] === 'undefined') {
            return new Promise(resolve => resolve([]));
        }

        const commentIds = parent.kids.slice(0, 1);

        for (const id of commentIds) {
            if (!cache.has(`comment:${id}`)) {
                // console.log(`getting comment ${id} from URL...`);

                const commentResp = await axios.get(`${process.env.HACKERNEWS_API_URL}/item/${id}.json`);
                const commentData: Record<string, any>[] = commentResp.data;

                cache.put(`comment:${id}`, commentData, 3600);
            }

            comments.push(cache.get(`comment:${id}`));
        }

        return new Promise(resolve => resolve(comments));
    };

    return commentsResolver;
}
