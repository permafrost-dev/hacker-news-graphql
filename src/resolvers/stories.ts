import axios from 'axios';

export function getResolver(topStoryIds, topStories, cache: any): any {
    const stories = async (_, { first, kind }) => {
        topStoryIds.splice(0, topStoryIds.length);
        topStories.splice(0, topStories.length);

        kind = `${kind}`.toLowerCase();

        const ids: number[] = [];
        const topstoryCacheKey = `${kind}storyids:${first}`;

        if (!cache.has(topstoryCacheKey)) {
            // console.log(`getting ${kind}storyids from URL...`);

            const resp = await axios.get(`${process.env.HACKERNEWS_API_URL}/${kind}stories.json?limitToFirst=${first}&orderBy="$key"`);
            const data: number[] = resp.data;

            cache.put(topstoryCacheKey, data, 60);
            ids.push(...data);
        }

        const cacheData = cache.get(topstoryCacheKey);

        ids.push(...cacheData);
        ids.splice(first);

        const storyDataPromises = ids
            .map(id => {
                if (cache.has(`${kind}story:${id}`)) {
                    return null;
                }

                // console.log(`getting story ${id} from URL...`);

                return axios.get(`${process.env.HACKERNEWS_API_URL}/item/${id}.json`);
            })
            .filter(item => item !== null);

        (await Promise.all(storyDataPromises))
            .map(resp => resp?.data)
            .filter(item => item !== null)
            .forEach(item => {
                cache.put(`${kind}story:${item.id}`, item, 600);
            });

        for (const id of ids) {
            topStories.push(cache.get(`${kind}story:${id}`));
        }

        return new Promise(resolve => resolve(topStories));
    };

    return stories;
}
