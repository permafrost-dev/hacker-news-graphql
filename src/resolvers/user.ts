/* eslint-disable no-unused-vars */

import { MemoryCache } from '../MemoryCache';
import axios from 'axios';

export function getResolver(cache: MemoryCache): any {
    const userResolver = async (_, { id }) => {
        if (cache.has(`user:${id}`)) {
            return cache.get(`user:${id}`);
        }

        const resp = await axios.get(`${process.env.HACKERNEWS_API_URL}/user/${id}.json`);

        cache.put(`user:${id}`, resp.data, 3600);

        return cache.get(`user:${id}`);
    };

    return userResolver;
}
