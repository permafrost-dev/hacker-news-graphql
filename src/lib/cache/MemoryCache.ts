import { Cache } from './Cache';

export interface CacheItem {
    ts: number;
    expires: number;
    value: any;
}

export interface CacheMap {
    [key: string]: CacheItem;
}

export class MemoryCache extends Cache {
    public map: CacheMap = {};

    constructor(keyPrefix = '') {
        super();

        if (keyPrefix.length) {
            keyPrefix += ':';
        }

        this.prefix = keyPrefix;
    }

    protected makeKey(key: string): string {
        return `${this.prefix}${key}`;
    }

    public async has(key: string): Promise<boolean> {
        this.purge();

        return new Promise(resolve => resolve(typeof this.map[this.makeKey(key)] !== 'undefined'));
    }

    public async get(key: string): Promise<any | null> {
        this.purge();

        const result = this.map[this.makeKey(key)]?.value ?? null;

        return new Promise(resolve => resolve(result));
    }

    public put(key: string, value: any, ttlSeconds: number) {
        this.map[this.makeKey(key)] = {
            ts: new Date().getTime(),
            expires: new Date().getTime() + ttlSeconds * 1000,
            value: value,
        };
    }

    public purge() {
        for (const prop in this.map) {
            if (new Date().getTime() >= this.map[prop].expires) {
                delete this.map[prop];
                continue;
            }
        }
    }

    public clear() {
        //
    }
}
