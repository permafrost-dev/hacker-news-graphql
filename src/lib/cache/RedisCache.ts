import { Cache } from './Cache';
const Redis = require('ioredis');

export class RedisCache extends Cache {
    public redis: typeof Redis;

    constructor(keyPrefix = '') {
        super();

        if (keyPrefix.length) {
            keyPrefix += ':';
        }

        this.prefix = keyPrefix;

        this.redis = new Redis(); //{ keyPrefix: keyPrefix });
    }

    public destroy() {
        this.redis.disconnect();
    }

    public async has(key: string): Promise<boolean> {
        const result = await this.redis.exists(key);

        return new Promise(resolve => resolve(result === 1));
    }

    public async get(key: string): Promise<any | null> {
        const result = await this.redis.get(key);

        return new Promise(resolve => resolve(JSON.parse(result)));
    }

    public put(key: string, value: any, ttlSeconds: number): void {
        this.redis.set(key, JSON.stringify(value), 'EX', ttlSeconds);
    }

    public purge() {
        //
    }

    public clear() {
        //
    }
}
