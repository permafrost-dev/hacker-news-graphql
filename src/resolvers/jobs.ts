import { Cache } from '@/lib/cache/Cache';
import axios from 'axios';
const hash = require('object-hash');

export function getResolver(storyIds, stories, cache: Cache): any {
    const result = async (_, { first, skipText }, { req }) => {
        const kind = 'job';

        storyIds.splice(0, storyIds.length);
        stories.splice(0, stories.length);

        const queryHash = hash(req.body);

        const ids: number[] = [];
        //const cacheKey = `jobstoryids:${first}-${skipText ? 'skipText' : 'dontSkipText'}`;
        const cacheKey = `jobstoryids:${first}:${queryHash}`;

        console.log('cacheKey = ', cacheKey);

        const hasKey = await cache.has(cacheKey);

        if (!hasKey) {
            console.log(`getting ${kind}storyids from URL...`);

            const resp = await axios.get(`${process.env.HACKERNEWS_API_URL}/jobstories.json?limitToFirst=${first + 5}&orderBy="$key"`);
            const data: number[] = resp.data;

            cache.put(cacheKey, data, 600);
            ids.push(...data);
        }

        const cacheData = await cache.get(cacheKey);

        ids.push(...cacheData);
        ids.splice(first);

        const storyDataPromises = ids
            .map(async id => {
                if (await cache.has(`${kind}story:${id}`)) {
                    return cache.get(`${kind}story:${id}`);
                }

                console.log(`getting story ${id} from URL...`);

                return axios.get(`${process.env.HACKERNEWS_API_URL}/item/${id}.json`);
            })
            .filter(item => item !== null);

        (await Promise.all(storyDataPromises))
            .map((resp: any) => resp?.data)
            .filter(item => item !== null)
            .filter(item => typeof item !== 'undefined')
            .forEach(item => {
                try {
                    cache.put(`${kind}story:${item.id}`, item, 3600);
                } catch (err) {
                    console.log(err);
                }
            });
        for (const id of ids) {
            const storyItem = await cache.get(`${kind}story:${id}`);

            if (!skipText || typeof storyItem['text'] === 'undefined' || storyItem['text'] === null) {
                stories.push(storyItem);
            }
        }

        return new Promise(resolve => resolve(stories));
    };

    return result;
}
