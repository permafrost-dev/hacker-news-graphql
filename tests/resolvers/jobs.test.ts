/* eslint-disable no-undef */

import { MemoryCache } from '@/lib/cache/MemoryCache';
import { getResolver } from '@/resolvers/jobs';
const hash = require('object-hash');

let cache: MemoryCache;
let resolver: any;
const storyIds: number[] = [];
const stories: Record<string, any>[] = [];

process.env.HACKERNEWS_API_URL = 'http://127.0.0.1:3031';

beforeEach(() => {
    cache = new MemoryCache();
    resolver = getResolver(storyIds, stories, cache);
});

it('it retrieves the story data and returns it', async () => {
    const req = { body: { first: 2, skipText: false } };
    const data = await resolver({}, { first: 2, skipText: false }, { req });

    for (const item of data) {
        item.expires = 0;
    }

    expect(data).toMatchSnapshot();
});

it('it retrieves the job ids and caches them', async () => {
    const req = { body: { query: 'abc', vars: 'bbb' } };
    const reqHash = hash(req.body);

    await resolver({}, { first: 2, skipText: false }, { req });

    expect(await cache.has(`jobstoryids:2:${reqHash}`)).toBeTruthy();
});

it('it retrieves each job story and caches it', async () => {
    const req = { body: { first: 2, skipText: false, a: 1, b: 2 } };
    const data = await resolver({}, { first: 2, skipText: false }, { req });

    data.forEach(item => {
        expect(cache.has(`jobstory:${item.id}`)).toBeTruthy();
    });

    for (const item in cache.map) {
        cache.map[item].ts = 0;
        cache.map[item].expires = 0;
    }

    expect(cache.map).toMatchSnapshot();
});
