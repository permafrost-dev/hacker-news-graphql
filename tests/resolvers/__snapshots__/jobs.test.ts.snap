// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it retrieves each job story and caches it 1`] = `
Object {
  "jobstory:27414470": Object {
    "expires": 0,
    "ts": 0,
    "value": Object {
      "by": "yiggydyang",
      "id": 27414470,
      "score": 1,
      "time": 1622998836,
      "title": "PocketSuite Is hiring Lead Engineers, Product Designers to enrich small business",
      "type": "job",
      "url": "https://www.ycombinator.com/companies/pocketsuite",
    },
  },
  "jobstory:27421523": Object {
    "expires": 0,
    "ts": 0,
    "value": Object {
      "by": "ifc",
      "id": 27421523,
      "score": 1,
      "time": 1623067272,
      "title": "Qventus (YC W15) Is Hiring Data Platform Engineers",
      "type": "job",
      "url": "https://jobs.lever.co/qventus/c0d604cf-ec1c-48ea-847d-5ffc7a67698d",
    },
  },
  "jobstoryids:2:16f0f9a19467c0d46c88080484e6fc4cafca7eb3": Object {
    "expires": 0,
    "ts": 0,
    "value": Array [
      27421523,
      27414470,
    ],
  },
}
`;

exports[`it retrieves the story data and returns it 1`] = `
Array [
  Object {
    "by": "ifc",
    "expires": 0,
    "id": 27421523,
    "score": 1,
    "time": 1623067272,
    "title": "Qventus (YC W15) Is Hiring Data Platform Engineers",
    "type": "job",
    "url": "https://jobs.lever.co/qventus/c0d604cf-ec1c-48ea-847d-5ffc7a67698d",
  },
  Object {
    "by": "yiggydyang",
    "expires": 0,
    "id": 27414470,
    "score": 1,
    "time": 1622998836,
    "title": "PocketSuite Is hiring Lead Engineers, Product Designers to enrich small business",
    "type": "job",
    "url": "https://www.ycombinator.com/companies/pocketsuite",
  },
]
`;
